Sources/Basics/TemporaryFile.swift

```diff
@@ -0,0 +1,122 @@
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    fileSystem: FileSystem = localFileSystem,
+    _ body: @escaping (AbsolutePath, @escaping (AbsolutePath) -> Void) async throws -> Result
+) throws -> Task<Result, Error> {
+    let temporaryDirectory = try createTemporaryDirectory(dir: dir, prefix: prefix, fileSystem: fileSystem)
+    
+    let task: Task<Result, Error> = Task {
+        try await withTaskCancellationHandler {
+            try await body(temporaryDirectory) { path in
+                try? fileSystem.removeFileTree(path)
+            }
+        } onCancel: {
+            try? fileSystem.removeFileTree(temporaryDirectory)
+        }
+        
+    }
+    
+    return task
+}
+
+/// Creates a temporary directory and evaluates a closure with the directory path as an argument.
+/// The temporary directory will live on disk while the closure is evaluated and will be deleted afterwards.
+///
+/// This function is basically a wrapper over posix's mkdtemp() function.
+///
+/// - Parameters:
+///     - dir: If specified the temporary directory will be created in this directory otherwise environment
+///            variables TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env
+///            variables are set, dir will be set to `/tmp/`.
+///     - prefix: The prefix to the temporary file name.
+///     - removeTreeOnDeinit: If enabled try to delete the whole directory tree otherwise remove only if its empty.
+///     - body: A closure to execute that receives the absolute path of the directory as an argument.
+///             If `body` has a return value, that value is also used as the
+///             return value for the `withTemporaryDirectory` function.
+///
+/// - Throws: `MakeDirectoryError` and rethrows all errors from `body`.
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public func withTemporaryDirectory<Result>(
+    dir: AbsolutePath? = nil,
+    prefix: String = "TemporaryDirectory",
+    fileSystem: FileSystem = localFileSystem,
+    removeTreeOnDeinit: Bool = false,
+    _ body: @escaping (AbsolutePath) async throws -> Result
+) throws -> Task<Result, Error> {
+    try withTemporaryDirectory(dir: dir, prefix: prefix, fileSystem: fileSystem) { path, cleanup in
+        defer { if removeTreeOnDeinit { cleanup(path) } }
+        return try await body(path)
+    }
+}
+
+private func createTemporaryDirectory(dir: AbsolutePath?, prefix: String, fileSystem: FileSystem) throws -> AbsolutePath {
+    // This random generation is needed so that
+    // it is more or less equal to generation using `mkdtemp` function
+    let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+    let randomSuffix = String((0..<6).map { _ in letters.randomElement()! })
+    
+    // Construct path to the temporary directory.
+    let templatePath = try AbsolutePath(prefix + ".\(randomSuffix)", relativeTo: determineTempDirectory(dir))
```

Comment:
since we have the `FileSystem` available now, we can probably also get rid of `determineTempDirectory` at this point... here is what it does:

```
public func determineTempDirectory(_ dir: AbsolutePath? = nil) throws -> AbsolutePath {
    let tmpDir = dir ?? localFileSystem.tempDirectory
    guard localFileSystem.isDirectory(tmpDir) else {
        throw TempFileError.couldNotFindTmpDir(tmpDir.pathString)
    }
    return tmpDir
}
```

Reply:
Done it!